WordPress/gutenberg
# Try: Extensibility: Enable hook-based Slot/Fill registration

Related: #1352

This pull request seeks to explore an pattern for plugin UI extensibility by exposing slots on a global hooks registry.

__Implementation notes:__

These changes make use of [`@wordpress/hooks@0.1.0-beta.5`](https://www.npmjs.com/package/@wordpress/hooks) to add a new hooks registry to the editor returned instance, available at `wp._wpGutenbergEditor.addFilter`. The behavior here of adding a hooks registry specific to editor (vs. a global registry) is not settled, but enabled as part of changes introduced to the hooks module in https://github.com/WordPress/packages/pull/40.

With these changes, a `registerSlot` action is monitored by a PluginSlots component after mounting. A plugin can ~fire this action to~ register a fill component by slot name. For example, entering the following in the browser developer tools console while running this branch will cause a message to be shown in the publish menu dropdown:

```js
_wpGutenbergEditor.registerFill( 'publish-dropdown', () => 'Hello!' );
```

A few alternative approaches were considered:

- Not leveraging hooks behavior and instead creating a separate API for registering fills, exposed by the editor, i.e. `editor.registerFill( name, FillComponent );`
   - It was cumbersome to try to maintain communication between the top-level API and the component responsible for rendering the fills. `PluginSlots` here can instead leverage hooks context provided by `HooksProvider`.
- Treating fills as a filter triggered during `PluginSlots` mount, i.e. instead of `this.state = { fills: {} };`, doing `this.state = { fills: applyFilters( 'registerFills', {} ) };`
   - Ensuring that plugins manipulate the object as expected is error-prone (must concatenate with or create a default array on slot name)
   - Requires that plugins register to this filter by the time the component mounts, meaning it is not possible to register after page load
- Maintaining Fills in Redux state
   - Possible, and not too dissimilar to current approach, but this would require storing component objects in Redux state, and was decided against for the goal of keeping Redux state as a plain object.

The included `HooksProvider` component (and `withHooks` higher-order component) enable a hooks registry scoped to a specific React tree, rather than on a single global. Therefore, each editor instance returned by `wp.editor.createEditorInstance` will have its own hooks registry (similar to having its own Redux store via `ReactReduxProvider`).

__To-Do:__

- [ ] Document added components
- [ ] Test added components

__Testing instructions:__

Enter the following into your browser developer tools console after loading the editor and verify that a "Hello!" message is shown when then toggling the Publish dropdown menu:

```js
_wpGutenbergEditor.doAction( 'registerFill', 'publish-dropdown', () => 'Hello!' );
```
